app/controllers/find/candidates/saved_courses_controller.rb

```diff
@@ -2,8 +2,35 @@ module Find
   module Candidates
     class SavedCoursesController < ApplicationController
       before_action :require_authentication
+      before_action :set_candidate
 
-      def index
+      def index; end
+
+      def create
+        course = Course.find(params[:course_id])
+
+        begin
+          @candidate.saved_courses.find_or_create_by!(course_id: course.id)
+          redirect_to find_course_path(provider_code: course.provider_code, course_code: course.course_code)
+        rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotSaved
+          redirect_to find_course_path(provider_code: course.provider_code, course_code: course.course_code), flash: { error: { message: t(".save_failed") } }
```

Comment:
Can we move to a service? `SaveCourseService`? 

if save_course_service.call
   redirect_to ...
else
   redirect_to
end

then you controll the exceptions in the service class. I'd prefer.

Reply:
I can see why is over abstraction, but I think we would want to:

1. Save the course
2. Control the flow in a simple way (true/false)
3. Send to sentry why it errored.

```ruby
class SaveCourseService
  def self.call(candidate:, course:)
    new(candidate, course).call
  end

  def initialize(candidate, course)
    @candidate = candidate
    @course = course
  end

  def call
    @candidate.saved_courses.find_or_create_by!(course_id: @course.id)
    true
  rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotSaved => e
    Sentry.capture_exception(e) # Notify Sentry here
    false
  end
end

```

Reply:
Also probably we will have to send events too to improve our metrics. This class will do that.